Kazoo: added basic kemi support

tao-communications opened the pull request:
<!-- Kamailio Pull Request Template -->

<!--
IMPORTANT:
  - for detailed contributing guidelines, read:
    https://github.com/kamailio/kamailio/blob/master/.github/CONTRIBUTING.md
  - pull requests must be done to master branch, unless they are backports
    of fixes from master branch to a stable branch
  - backports to stable branches must be done with 'git cherry-pick -x ...'
  - code is contributed under BSD for core and main components (tm, sl, auth, tls)
  - code is contributed GPLv2 or a compatible license for the other components
  - GPL code is contributed with OpenSSL licensing exception
-->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask on sr-dev mailing list -->
- [*] Commit message has the format required by CONTRIBUTING guide
- [*] Commits are split per component (core, individual modules, libs, utils, ...)
- [*] Each component has a single commit (if not, squash them into one commit)
- [*] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [*] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply -->
- [ ] PR should be backported to stable branches
- [*] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description
<!-- Describe your changes in detail -->
Added kemi support for
Parameter:
- event_callback
Functions:
- kazoo_subscribe(json_description)  
- kazoo_publish
Documentation:
- added section for event_callback parameter

miconda commented:
Thanks, really appreciated being a module I am not using much, so I haven't approach it from KEMI perspective yet!

@lazedo - can you quickly check this PR and see if there is something you want to comment on?

Otherwise, I am fine to merge.

lazedo commented:
@miconda i'll try to take a deeper look next week. at first glance,  some formatting issues and use of route_no which can break things already in place.

lazedo commented on src/modules/kazoo/kazoo.c:
```diff
@@ -273,6 +279,12 @@ static int mod_init(void) {
    		return -1;
     }
 
+    if(faked_msg_init()<0)
+    {
+        LM_ERR("failed to init kazoo local sip msg\n");
+        return -1;
+    }
+
```
why was this added ?

tao-communications replied:
This has been removed.

lazedo commented on src/modules/kazoo/kazoo.c:
```diff
@@ -312,7 +324,30 @@ static int mod_init(void) {
     }
 
-
+    sr_kemi_eng_t *keng = NULL;
+	  if(kazoo_event_callback.s!=NULL && kazoo_event_callback.len>0) {
+		  keng = sr_kemi_eng_get();
+		  if(keng==NULL) {
+			  LM_ERR("failed to find kemi engine\n");
+			  return -1;
+		  }
+		  kazoo_route_no=-1;
+	  } 
+    else {
+		  route_no=route_lookup(&event_rt, "kazoo:consumer-event");
+		  if (route_no==-1)
+		  {
+			  LM_ERR("failed to find event_route[kazoo:consumer-event]\n");
+			  return -1;
+		  }
+		  if (event_rt.rlist[route_no]==0)
+		  {
+			  LM_WARN("event_route[kazoo:consumer-event] is empty\n");
+		  }
+		  kazoo_route_no=route_no;
+    }
+    
+    
```
please use a different logic here. `kazoo:consumer-event` is not mandatory. you could set a variable like `kz_use_kemi` that is set if `kazoo_event_callback` is set and `kemi` is valid. then use that variable when needed

tao-communications replied:
Now removed route_lookup for kazoo:consumer-event as it may or may not exist.

lazedo commented on src/modules/kazoo/kazoo.c:
```diff
@@ -435,27 +487,46 @@ static int fire_init_event(int rank)
-		{
-			LM_ERR("exit due to 'drop' in event route\n");
-			return -1;
+	if (kazoo_route_no>=0) {
+		rt = route_get(&event_rt, "kazoo:mod-init");
+        if(rt>=0 && event_rt.rlist[rt]!=NULL) {
+			LM_DBG("executing event_route[kazoo:mod-init] (%d)\n", rt);
```
formatting looks odd

tao-communications replied:
Now fixed formatting.

lazedo commented on src/modules/kazoo/kz_amqp.c:
```diff
@@ -2362,30 +2378,48 @@ void kz_amqp_consumer_event(kz_amqp_consumer_delivery_ptr Evt)
-                sprintf(buffer, "kazoo:consumer-event-%s", key);
+                sprintf(buffer, "kazoo:consumer-event-%s-%s", key, subkey);
                 for (p=buffer ; *p; ++p) *p = tolower(*p);
                 for (p=buffer ; *p; ++p) if(*p == '_') *p = '-';
-				if(kz_amqp_consumer_fire_event(buffer) != 0) {
-					sprintf(buffer, "kazoo:consumer-event");
-					if(kz_amqp_consumer_fire_event(buffer) != 0) {
-						LM_ERR("kazoo:consumer-event not found\n");
-					}
-				}
+                if(kz_amqp_consumer_fire_event(buffer) != 0) {
+                    sprintf(buffer, "kazoo:consumer-event-%s", key);
+                    for (p=buffer ; *p; ++p) *p = tolower(*p);
+                    for (p=buffer ; *p; ++p) if(*p == '_') *p = '-';
+                    if(kz_amqp_consumer_fire_event(buffer) != 0) {
+                        sprintf(buffer, "kazoo:consumer-event");
+                        if(kz_amqp_consumer_fire_event(buffer) != 0) {
+                            LM_ERR("kazoo:consumer-event not found\n");
+                        }
+                    }
+                }
             }
         }
+    } else {
+        keng = sr_kemi_eng_get();
+        if(keng!=NULL) {
+            sip_msg_t *msg;
+            str evrtname = str_init("kazoo:consumer-event");
+
+            rtb = get_route_type();
+            msg = faked_msg_next();
+            if(sr_kemi_route(keng, msg, EVENT_ROUTE, &kazoo_event_callback, &evrtname)<0)                {
+                LM_ERR("error running event route kemi callback\n");
+            }
+            set_route_type(rtb);
+        } else {
+            LM_ERR("no event route or kemi callback found for execution\n");
+        }
     }
+
```
split this in two different functions that are evoked depending on `kemi` being active

tao-communications replied:
Now depending on using cfg or kemi:
     - kz_amqp_consumer_event() calls kz_amqp_consumer_event_cfg() or kz_amqp_consumer_event_kemi()

lazedo commented on src/modules/kazoo/kazoo.c:
```diff
@@ -435,27 +487,46 @@ static int fire_init_event(int rank)
-		set_route_type(REQUEST_ROUTE);
-		init_run_actions_ctx(&ctx);
-		run_top_route(event_rt.rlist[rt], fmsg, &ctx);
-		if(ctx.run_flags&DROP_R_F)
-		{
-			LM_ERR("exit due to 'drop' in event route\n");
-			return -1;
+	if (kazoo_route_no>=0) {
+		rt = route_get(&event_rt, "kazoo:mod-init");
+        if(rt>=0 && event_rt.rlist[rt]!=NULL) {
+			LM_DBG("executing event_route[kazoo:mod-init] (%d)\n", rt);
+			if(faked_msg_init()<0)
+				return -1;
+			fmsg = faked_msg_next();
+			rtb = get_route_type();
+			set_route_type(REQUEST_ROUTE);
+			init_run_actions_ctx(&ctx);
+			run_top_route(event_rt.rlist[rt], fmsg, &ctx);
+			if(ctx.run_flags&DROP_R_F)
+			{
+				LM_ERR("exit due to 'drop' in event route\n");
+				return -1;
+			}
+			set_route_type(rtb);
+		}
+	} 
+  else {
+		keng = sr_kemi_eng_get();
+		if(keng!=NULL) {
+			sip_msg_t *msg;
+			str evrtname = str_init("kazoo:mod-init");
+			rtb = get_route_type();
+			msg = faked_msg_next();
+			if(sr_kemi_route(keng, msg, EVENT_ROUTE, &kazoo_event_callback, &evrtname)<0) {
+				LM_ERR("error running event route kemi callback\n");
+		  }
+		  set_route_type(rtb);
+		} 
+    else {
+			LM_ERR("no event route or kemi callback found for execution\n");
```
split this in two different function that are evoked by `fire_init_event` depending on `kemi` being active

tao-communications replied:
Now depending on using cfg or kemi:
     - fire_init_event() calls fire_init_event_cfg() or fire_init_event_kemi()

tao-communications commented:
I have addressed the suggested changes in the new commit:
    - removed route_lookup for kazoo:consumer-event as it may or may not exist
    - depending on using cfg or kemi:
     - fire_init_event() calls fire_init_event_cfg() or fire_init_event_kemi()
     - kz_amqp_consumer_event() calls kz_amqp_consumer_event_cfg() or
and tidied up formatting.

Tested on my test servers for: kazoo_subscribe,  kazoo_publish, kazoo:consumer-event, with the traditional cfg and Kemi (Lua).
    kz_amqp_consumer_event_kemi()

lazedo commented:
@tao-communications great job, thank you. i do have a final request, when that's done, please squash your commits into a single commit, `kazoo: add basic kemi support`

i feel that `kazoo_route_no` doesn't express well the intent for the variable which is to use `kemi` instead of `cfg`.

can you please rename `kazoo_route_no`  to `kazoo_kemi_enabled` with default 0 and set it to 1 in  [here](https://github.com/kamailio/kamailio/pull/2324/files#diff-52d37ad2407d869cdcee8ffc29166e62R330)
and then use 
```
if ( kazoo_kemi_enabled )  {
   do kemi routine
} else {
   do cfg routine
}
```
where applicable, for example in `fire_init_event`
```
	if (kazoo_kemi_enabled) {
		return fire_init_event_kemi();
	}
	else {
		return fire_init_event_cfg();
	}
``

tao-communications commented:
Thanks for reviewing! I made a new commit to address these.
Please let me know if there are further things that need doing.

lazedo commented:
@tao-communications all good. please squash your commits so it can be merged

miconda commented:
@lazedo: @tao-communications does not have commit access yet.

You can squash all commits via github merge button, just click on arrow down next to `Merge pull request` and choose `Squash and merge`. It is no need to have the squash done elsewhere.

I can do it if you prefer.

lazedo commented:
@miconda i know that, i just asked to squash 6 commits into one. waiting on this and i will `Rebase and merge`. i want to preserve the original commit and not do a `merge`. thank you

lazedo commented:
@miconda why did you dismissed my review ?

miconda commented:
@lazedo - I dismissed the review because it was requesting an external squash, which is not necessary in my opinion. You can set the any commit message when you squash and merge via github.com web UI.

tao-communications commented:
I have attempted a git squash and hopefully it's OK.

lazedo commented:
@tao-communications looks good, just waiting for travis to complete. thanks.

lazedo commented:
@tao-communications thank you again for your contribution.

tao-communications commented:
Thank you very much @lazedo and @miconda for the review and support!
